sklearn/cross_validation.py

```diff
@@ -506,26 +563,40 @@ class LeaveOneLabelOut(_PartitionIterator):
 
     """
 
-    def __init__(self, labels, indices=None):
-        super(LeaveOneLabelOut, self).__init__(len(labels), indices)
-        # We make a copy of labels to avoid side-effects during iteration
-        self.labels = np.array(labels, copy=True)
-        self.unique_labels = np.unique(labels)
-        self.n_unique_labels = len(self.unique_labels)
+    def __init__(self, labels=None, indices=None):
+        n = _num_samples(labels) if labels is not None else None
+        super(LeaveOneLabelOut, self).__init__(n, indices)
+        self.labels = labels
 
-    def _iter_test_masks(self):
-        for i in self.unique_labels:
-            yield self.labels == i
+    def _iter_test_masks(self, y):
+        y = self.labels if y is None else y
+        # We make a copy of labels to avoid side-effects during iteration
+        y = np.array(y, copy=True)
+        for i in np.unique(y):
+            yield y == i
 
     def __repr__(self):
-        return '%s.%s(labels=%s)' % (
+        format_vars = (
             self.__class__.__module__,
             self.__class__.__name__,
             self.labels,
         )
+        if self.labels is None:
+            return "{0}.{1}(labels={2})".format(*format_vars)
+        return "{0}.{1}()".format(*format_vars)
 
-    def __len__(self):
-        return self.n_unique_labels
+    @property
+    def unique_labels(self):
+        _deprecate_label_attribute(self, "unique_labels")
+        return np.unique(self.labels)
+
+    @property
+    def n_unique_labels(self):
+        _deprecate_label_attribute(self, "n_unique_labels")
+        return len(np.unique(self.labels))
```

Comment:
This is not so much about initializing in ``__init__`` (which wouldn't be a problem here since these are not estimators) than warning only of someone actually uses them. But you did the right thing.